File: src/main/java/algo/sort/RankTeamsByVotes_1366.java
package algo.sort;

import java.util.HashMap;
import java.util.PriorityQueue;

public class RankTeamsByVotes_1366 {
    public static String rankTeams(String[] votes) {
        var numTeams = votes[0].length();
        var count = new HashMap<String, int[]>();
        var pq = new PriorityQueue<Character>((a, b) -> {
            var va = count.get(a + "");
            var vb = count.get(b + "");
            for (var i = 0; i < numTeams; i++) {
                if (va[i] > vb[i]) {
                    return -1;
                } else if (va[i] < vb[i]) {
                    return 1;
                }
            }
            return a - b;
        });
        for (var v : votes) {
            for (var i = 0; i < numTeams; i++) {
                var team = v.charAt(i) + "";
                if (count.containsKey(team)) {
                    count.get(team)[i]++;
                } else {
                    var numVote = new int[numTeams];
                    numVote[i]++;
                    count.put(team, numVote);
                }
            }
        }
        for (var key : count.keySet()) {
            pq.offer(key.charAt(0));
        }

        var result = "";
        while (!pq.isEmpty()) {
            result += pq.poll();
        }

        return result;
    }

    public static void main(String[] args) {
        System.out.println(rankTeams(new String[]{"WXYZ", "XYZW"}));
        System.out.println(rankTeams(new String[]{"ABC","ACB","ABC","ACB","ACB"}));
    }
}

File: src/main/kotlin/algo/sort/RankTeamsByVotes_1366.kt
package algo.sort

import java.util.PriorityQueue

class KtRankTeamsByVotes_1366 {

    fun rankTeams(votes: Array<String>): String {
        // collect votes for each team
        val voteCount: MutableMap<Char, IntArray> = HashMap(26)
        votes.forEach { v ->
            v.indices.forEach { i ->
                voteCount.computeIfAbsent(v[i]) { IntArray(v.length) }
                voteCount[v[i]]!![i] = voteCount[v[i]]!![i] + 1
            }
        }

        val teams = voteCount.keys
        return teams.sortedWith(object: Comparator<Char> {
            override fun compare(a: Char?, b: Char?): Int {
                val va = voteCount[a]!!
                val vb = voteCount[b]!!
                for (i in va.indices) {
                    if (va[i] != vb[i]) {
                        return vb[i] - va[i]
                    }
                }
                return 0
            }
        }).joinToString("")
    }
}

fun main() {
    val rank = KtRankTeamsByVotes_1366()
    println(rank.rankTeams(arrayOf("ABC","ACB","ABC","ACB","ACB")))
    println(rank.rankTeams(arrayOf("ZMNAGUEDSJYLBOPHRQICWFXTVK")))
    println(rank.rankTeams(arrayOf("BCA","CAB","CBA","ABC","ACB","BAC")))
}